### ISSUE TITLE
Using `async` compile function in addExtension causes "this.defaultRenderer is not a function" error

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#2217]
Reported by @sjml here: https://github.com/11ty/11ty-website/pull/1286#issuecomment-1016445009

Broken and working versions:

```diff
module.exports = function(eleventyConfig) {
  eleventyConfig.addExtension("md", {
-    compile: async function (inputContent, inputPath) {
+    compile: function (inputContent, inputPath) {
      return async function(data) {
        return this.defaultRenderer(data);
      };
    }
  });
};
```

**Comment by AramZS** [comment]
Was just about to file this bug, looking forward to resolution! Glad someone else hit it.

**Comment by AramZS** [comment]
Actually, the bug I hit (`defaultRenderer` not being on `this`) was with: https://www.11ty.dev/docs/languages/custom/#overriding-an-existing-template-language

I think it needs to be corrected in that documentation as well?

**Comment by zachleat** [comment]
There were some missing test cases here, specifically with `async` compile and `async` render (and using an arrow function for `render` or `compile`). I think we want `defaultRenderer` available in as many of these cases as we can (though it won’t work when both `render` and `compile` are arrow functions, because it needs at least one `this` to bind to).

Those fixes will ship with 1.0.1. Thanks y’all!
